### DISCUSSION TITLE
Consider renaming `createSSGHelpers` to `createSSRHelpers`

### DISCUSSION THREAD

**Discussion by kripod** [trpc/trpc#370] [Ideas]
The `createSSGHelpers` method of `@trpc/react` can be used for both SSG and SSR, and the former is just a special form of prerendering – "ahead-of-time SSR".

Also, I propose moving `@trpc/react/ssg` to `@trpc/react/ssr`.

**Comment by KATT** [comment]
It is intentional as I personally don't see much use of using those helpers for SSR rather than SSG since you can just turn on `{ ssr: true }`, I've never liked `getServerSideProps` myself as it always adds a delay to the route change.

Did you know that even with SSR you can take advantage of Vercel's edge cache?

https://trpc.io/docs/caching#app-caching

Read the twitter thread that's linked there. As long as each dynamic page has been visited at least once since the last deploy and you have cache headers you are able to get it statically delivered.

**Reply by kripod** [reply]
Thank you for mentioning. I’ve read the entire docs before starting to use the library and saw that particular tweet, too. It’s a great idea which has even been hinted on by the developers behind [Remix](https://remix.run/).

I’m not sure if the usage of `getInitialProps` avoids the route change delay, though. I thought it would work similarly to gSSP, but haven’t considered if links could be prefetched without gSSP but with a `getInitialProps` present.

I’m not sure if migrating away from gSSP would be possible whenever my company decides to make some pages password-protectable.

Thank you for explaining your thoughts in great detail – now I understand the decision behind naming the method after "SSG".

**Reply by KATT** [reply]
> I’m not sure if the usage of `getInitialProps` avoids the route change delay, though

It seems to according to my tests!


> I’m not sure if migrating away from gSSP would be possible whenever my company decides to make some pages password-protectable.

Same in my case. I use an auth cookie to auth the requests so it works fine with ssr. If I did deploy on vercel I'd just make sure not to send cache headers whenever the auth cookie is present. Had a similar conversation with @cyrus-za the other day.

I.e like the below + check that `ctx.req` doesn't include the auth cookie.

![image](https://user-images.githubusercontent.com/459267/118403106-25650500-b66d-11eb-953b-994e120d3826.png)

**Reply by kripod** [reply]
Whoa, that’s a game changer! So it turns out that pages get prefetched by Next’s `<Link>` component even when the entire app has a `getInitialProps` method attached? 😲

That opens up many new possibilities when combined with the `stale-while-revalidate` caching technique you’ve just mentioned.

I would appreciate if you could share some thoughts about password-protecting pages as efficiently as possible – if I understood correctly that you’re working on implementing something similar.

**Reply by KATT** [reply]
Yes, I am, I won't take any accountability for your app's security, but my prototyping so far seems to check out. Try it out!

**Reply by bennettdams** [reply]
I think the question is still relevant, but maybe an even more generic name makes sense: "prefetch helpers", as they can be used for both `getStaticProps` and `getServerSideProps`. I would be down to create a PR for such changes - what do you think? 

Why I think this is important: I personally am not brave enough to use `ssr: true` with `getInitialProps` and I think others are not as well. I can just prefetch some things on the server with the helpers, and others are just fetched on the client. Just because some people use `ssr: true` does not mean that the helpers can only be used for SSG, which means their current name does not represent their functionality.

Btw: Does someone know whether we will be able to prefetch queries inside a React Server Component? If yes, a more general name like "prefetch helpers" makes way more sense.
